storage/Devices/MsdosImpl.cc

```diff
@@ -167,6 +168,84 @@ namespace storage
     }
 
 
+    const Partition*
+    Msdos::Impl::get_extended() const
+    {
+	vector<const Partition*> partitions = PartitionTable::Impl::get_partitions();
+	for (const Partition* partition : partitions)
+	{
+	    if (partition->get_type() == PartitionType::EXTENDED)
+		return partition;
+	}
+
+	ST_THROW(Exception("has no extended partition"));
+    }
+
+
+    vector<Partition*>
+    Msdos::Impl::get_partitions()
+    {
+	vector<Partition*> partitions = PartitionTable::Impl::get_partitions();
+
+	for (Partition* partition : partitions)
+	{
+	    if (partition->get_type() == PartitionType::EXTENDED)
+	    {
+		vector<Partition*> logicals = get_logical_partitions(partition);
+		partitions.insert(partitions.end(), logicals.begin(), logicals.end());
+		break;
+	    }
+	}
+
+	return partitions;
+    }
+
+
+    vector<const Partition*>
+    Msdos::Impl::get_partitions() const
+    {
+	vector<const Partition*> partitions = PartitionTable::Impl::get_partitions();
+
+	for (const Partition* partition : partitions)
+	{
+	    if (partition->get_type() == PartitionType::EXTENDED)
+	    {
+		vector<const Partition*> logicals = get_logical_partitions(partition);
+		partitions.insert(partitions.end(), logicals.begin(), logicals.end());
+		break;
+	    }
+	}
+
+	return partitions;
+    }
+
+
```

Comment:
Just asking from my C++ ignorance. There is no easy way to avoid the code duplication between `vector<Partition*> Msdos::Impl::get_partitions()` and `vector<const Partition*> Msdos::Impl::get_partitions() const`?

Same goes for `get_logical_partitions`